compiler/rustc_mir_build/src/build/matches/mod.rs

```diff
@@ -1800,7 +1811,14 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         let otherwise_post_guard_block = otherwise_candidate.pre_binding_block.unwrap();
         self.break_for_else(otherwise_post_guard_block, else_target, self.source_info(expr_span));
 
-        self.declare_bindings(None, pat.span.to(span), pat, ArmHasGuard(false), opt_expr_place);
+        self.declare_bindings(
```

Comment:
Another option is to systematically set `self.source_scope` to the scope returned here, but I'm hesitant to do that as this would also affect the path that normal `let` bindings flow through.